MetalCypher/MetalCypher.metal

```diff
@@ -11,33 +11,47 @@
 
 using namespace metal;
 
-kernel void bruteForce(constant byte const  * hash      [[buffer(BruteForceParameterHash)]],
-                       constant uint const  * input     [[buffer(BruteForceParameterInput)]],
-                       device   uint        * match     [[buffer(BruteForceParameterMatch)]],
+kernel void bruteForce(constant uint4 const * hash      [[buffer(BruteForceParameterHash)]],
+                       constant uint  const * input     [[buffer(BruteForceParameterInput)]],
+                       device   atomic_uint * match     [[buffer(BruteForceParameterMatch)]],
                        thread   uint          threadID  [[thread_position_in_grid]]) {
+  if (atomic_load_explicit(match, memory_order_relaxed) > 0) return;
+  
   thread byte password[MAX_PASSWORD_LENGTH];
-  uint passwordSize;
+  thread uint passwordSize;
   
   passwordFrom(*input + threadID, password, &passwordSize);
-  MD5 md5(password, passwordSize);
+  thread MD5 md5(password, passwordSize);
+
+  thread bool isNotAMatch = (hash->x ^ md5.output.x) | (hash->y ^ md5.output.y) | (hash->z ^ md5.output.z) | (hash->w ^ md5.output.w);
   
-  for (uint i = 0; i < HASH_SIZE; i++) {
-    if (hash[i] != md5.output[i]) return;
+  thread uint mask = 0;
+  thread size_t uintSizeInBits = sizeof(uint) * BYTE_SIZE_IN_BITS;
+  for (thread uint i = 0; i < uintSizeInBits; i++) {
+    mask |= isNotAMatch << i;
   }
-  *match = threadID + 1;
+  mask = ~mask;
+  
+  thread uint store = mask & (threadID + 1);
```

Comment:
I'm curious, what does the `thread` keyword do exactly?

Reply:
This is called by Metal the Address Space.

In Metal, every variable, especially pointers, may be part of an Address Space. The address space comes as a prefix before the variable.
`device` - the variable can be read or written from any thread.
`constant` - is just like `device`, but is read-only.
`threadgroup` - the variable can be read or written by the threads within a thread pool, or a threadgroup.
`thread` - the variable can be read or written within the specific thread only. The advantage of a thread variable is that it doesn't require atomicity, since this variable can't be edited from other threads.

The default for a variable is to be of the `thread` address space. I am being verbose to make it clear that that variable isn't shared among other threads at all. FYI pointer variables are required  by the compiler to have the address space defined, while simple variables are not. I added the word `thread` also for the sake of being consistent. Everything has the address space.

As a curiosity, you can see that this is not true for the method parameter `match`, that you can see that it is of the `device` address space. This parameter has to be that way, because this is the output to the CPU, that indicates that a match was found, and it says which is the password index that is the match. You can see that it has to be atomically set, otherwise another thread could set it to 0 after the match was found. But it is expensive to the processor. `thread` variables don't have this cost.

Reply:
Hmm, very interesting... Thanks for the clarification man :)